LedgerSMB/Database/Loadorder.pm

```diff
@@ -0,0 +1,176 @@
+=head1 NAME
+
+LedgerSMB::Database::Loadorder - LOADORDER parsing
+
+=cut
+
+package LedgerSMB::Database::Loadorder;
+use strict;
+use warnings;
+
+use LedgerSMB::Database::Change;
+use Cwd;
+
+=head1 SYNOPSIS
+
+my $loadorder = LedgerSMB::Database::Loadorder->new('path/to/loadorder');
+my @scripts = $loadorder->scripts;
+
+# to make an index of scripts and their sha hashes for db revision checking:
+
+$loadorder->makeindex()
+
+But see the notes about locking below
+
+=head1 METHODS
+
+=head2 constructor
+
+LedgerSMB::Database::Loadorder->new($path);
+
+=cut
+
+sub new {
+    my ($package, $path) = @_;
+    bless {_path => $path }, $package;
+}
+
+=head2 scripts
+
+Returns a list of LedgerSMB::Database::Change objects
+
+=cut
+
+my $reload_subsequent;
+sub scripts {
+    my ($self) = @_;
+    return @{$self->{_scripts}} if $self->{_scripts};
+    my $loadorder;
+    local $!;
+    local $@;
+    open(LOAD, '<', $self->{_path}) or
+        die 'FileError on ' . Cwd::abs_path($self->{_path}) . ": $!";
+    $reload_subsequent = 0;
+    my @scripts =
+       map { $self->_process_script($_)}
+       grep { $_ =~ /\S/ }
+       map { my $string = $_; $string =~ s/#.*$//; $string }
+       <LOAD>;
+    close LOAD;
+    $self->{_scripts} = \@scripts;
+    $reload_subsequent = 0;
+    return @scripts;
+}
+
+sub _process_script {
+    my ($self, $line) = @_;
+    chomp($line);
+    my $sigil = '';
+    if ($line =~ /^([!^]+)/){
+        $sigil = $1 if $1;
+        $line =~ s/^\Q$sigil\E//;
+    }
+    $reload_subsequent ||= ( $sigil =~ /\Q^\E/ );
+    my $no_transactions = ( $sigil =~ /\Q!\E/ );
+    return LedgerSMB::Database::Change->new(
+        $self->path($line),
+        {
+            reload_subsequent => $reload_subsequent,
+            no_transactions => $no_transactions
+        },
+    );
+}
+
+=head2 makeindex
+
+Creates an index of the files at $path/LOADORDER.idx and locks the file.
+
+If the file is already locks throws an error "LockError"
+
+The LOADORDER.idx remains empty.
+
+=cut
+
+sub makeindex {
+    my ($self) = @_;
+    die 'LockError' if -f $self->path('LOADORDER.idx');
+    open TEMP, '>', $self->path('LOADORDER.idx');
+    $self->{_locked} = 1;
+    for my $script ($self->scripts){
+        $script->load_contents;
+    }
+    close TEMP;
```

Comment:
Hmmm. If we leave things like this and we crash hard elsewhere, our patch directory remains locked forever. Can't we remove the lock file in something like a finally clause in the same routine where we open/create it?